pkg/rmb/rmb.go

```diff
@@ -0,0 +1,251 @@
+package rmb
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"time"
+
+	"github.com/gomodule/redigo/redis"
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog/log"
+)
+
+const (
+	systemLocalBus = "msgbus.system.local"
+	replyBus       = "msgbus.system.reply"
+	numWorkers     = 5
+)
+
+type twinKeyID struct{}
+type messageKey struct{}
+
+// Message is an struct used to communicate over the messagebus
+type Message struct {
+	Version    int      `json:"ver"`
+	UID        string   `json:"uid"`
+	Command    string   `json:"cmd"`
+	Expiration int      `json:"exp"`
+	Retry      int      `json:"try"`
+	Data       string   `json:"dat"`
+	TwinSrc    uint32   `json:"src"`
+	TwinDest   []uint32 `json:"dest"`
+	Retqueue   string   `json:"ret"`
+	Schema     string   `json:"shm"`
+	Epoch      int64    `json:"now"`
+	Err        string   `json:"err"`
+}
+
+// MessageBus is a struct that contains everything required to run the message bus
+type MessageBus struct {
+	Context  context.Context
+	pool     *redis.Pool
+	handlers map[string]func(ctx context.Context, payload []byte) (interface{}, error)
+}
+
+// New creates a new message bus
+func New(ctx context.Context, address string) (*MessageBus, error) {
+	pool, err := newRedisPool(address)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to connect to %s", address)
+	}
+
+	return &MessageBus{
+		pool:     pool,
+		Context:  ctx,
+		handlers: make(map[string]func(ctx context.Context, payload []byte) (interface{}, error)),
+	}, nil
+}
+
+// WithHandler adds a topic handler to the messagebus
+func (m *MessageBus) WithHandler(topic string, handler func(ctx context.Context, payload []byte) (interface{}, error)) {
+	m.handlers[topic] = handler
+}
+
+// Run runs listeners to the configured handlers
+// and will trigger the handlers in the case an event comes in
+func (m *MessageBus) Run() error {
+	con := m.pool.Get()
+	defer con.Close()
+
+	topics := make([]string, len(m.handlers))
+	for topic := range m.handlers {
+		topics = append(topics, topic)
+	}
+
+	jobs := make(chan Message, numWorkers)
+	for i := 1; i <= numWorkers; i++ {
+		go m.worker(context.Background(), jobs)
+	}
+
+	for {
+		if m.Context.Err() != nil {
+			return nil
+		}
+
+		data, err := redis.ByteSlices(con.Do("BLPOP", redis.Args{}.AddFlat(topics).Add(0)...))
+		if err != nil {
+			log.Err(err).Msg("failed to read from system local messagebus")
+			return err
+		}
+
+		var message Message
+		err = json.Unmarshal(data[1], &message)
+		if err != nil {
+			log.Err(err).Msg("failed to unmarshal message")
+			continue
+		}
+
+		_, ok := m.handlers[string(data[0])]
+		if !ok {
+			log.Debug().Msg("handler not found")
+			continue
+		}
+
+		jobs <- message
```

Comment:
while this should work but there is possible loss of messages that can happen if a message is popped from the queue but then this code here was stopped before the message get chance to be processed. I am not sure if this is something that we should handle ourselves or by the rmb itself! @maxux ?